### PR TITLE
Tp, Paintgun, refractor rainbow controller, refractor raycast

### PR DESCRIPTION
**Pull request by TheGreenBandit**
rainbow needs to be cleaned up

### CODE REVIEW

File: src/util/color.hpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+namespace big
+{
+	inline class rgb
+	{
+	public:
+		float red   = 255;
+		float green = 0;
+		float blue  = 0;
+	} rgb{};
+}
```
Comment (Yimura) [inline src/util/color.hpp]: Why are you saving data under util?

File: src/backend/looped/weapons/tpgun.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+			if (PAD::IS_DISABLED_CONTROL_PRESSED(0, (int)ControllerInputs::INPUT_AIM))
+			{
+				if (PAD::IS_DISABLED_CONTROL_JUST_RELEASED(0, (int)ControllerInputs::INPUT_ATTACK))
+				{
+					ENTITY::SET_ENTITY_COORDS(self::ped, c.x, c.y, c.z, 0, 0, 0, 0);
+				}
+			}
+		}
+	}
+}
```
Comment (Yimura) [inline src/backend/looped/weapons/tpgun.cpp:23]: This can be made into a looped command.

File: src/backend/looped/weapons/paintgun.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+					    g.weapons.paintgun.col[3],
+					    -1,
+					    true,
+					    false,
+					    false);
+				}
+			}
+		}
+	}
+}
```
Comment (Yimura) [inline src/backend/looped/weapons/paintgun.cpp:50]: This can be made into a looped command.

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 4]
```diff
@@ -39,96 +28,36 @@ namespace big
...
 			{
-				VEHICLE::SET_VEHICLE_TYRE_SMOKE_COLOR(vehicle, red, green, blue);
+				VEHICLE::SET_VEHICLE_TYRE_SMOKE_COLOR(self::veh, red, green, blue);
 			}
 
 			delay = std::chrono::milliseconds(((delay_step * 10) + 10) - (g.vehicle.rainbow_paint.speed * delay_step));
 			last_rgb_run_time = std::chrono::system_clock::now();
 		}
-	}
+	};
```
Comment (Yimura) [inline src/backend/looped/vehicle/rgb_paint.cpp]: Random semicolon

File: src/views/self/view_weapons.cpp [hunk 5]
```diff
@@ -216,12 +220,12 @@ namespace big
 			ImGui::PushItemWidth(250);
 			if (ImGui::BeginCombo("Attachments", selected_weapon_attachment.c_str()))
 			{
-				auto weapon = g_gta_data_service->weapon_by_hash(selected_weapon_hash);
+				weapon_item weapon = g_gta_data_service->weapon_by_hash(selected_weapon_hash);
```
Comment (Yimura) [inline src/views/self/view_weapons.cpp:222]: Any reason for changing the type from `auto` to `weapon_item`?

File: src/views/self/view_weapons.cpp [hunk 6]
```diff
@@ -216,12 +220,12 @@ namespace big
...
 				if (!weapon.m_attachments.empty())
 				{
 					for (std::string attachment : weapon.m_attachments)
 					{
-						auto attachment_component   = g_gta_data_service->weapon_component_by_name(attachment);
+						weapon_component attachment_component   = g_gta_data_service->weapon_component_by_name(attachment);
```
Comment (Yimura) [inline src/views/self/view_weapons.cpp:227]: Any reason for changing the type from `auto` to `weapon_item`?

File: src/views/self/view_weapons.cpp [hunk 7]
```diff
@@ -330,7 +334,7 @@ namespace big
 				for (auto& weapon_hash : g.weapons.weapon_hotkeys[selected_key])
 				{
 					ImGui::PushID(counter);
-					auto weapon = g_gta_data_service->weapon_by_hash(weapon_hash);
+					weapon_item weapon = g_gta_data_service->weapon_by_hash(weapon_hash);
```
Comment (Yimura) [inline src/views/self/view_weapons.cpp:336]: Any reason for changing the type from `auto` to `weapon_item`?

File: src/views/self/view_weapons.cpp [hunk 8]
```diff
@@ -280,7 +284,7 @@ namespace big
 			ImGui::PushItemWidth(250);
 			if (ImGui::BeginListBox("Saved Loadouts", ImVec2(200, 200)))
 			{
-				for (auto filename : persist_weapons::list_weapon_loadouts())
+				for (std::string filename : persist_weapons::list_weapon_loadouts())
```
Comment (Yimura) [inline src/views/self/view_weapons.cpp:286]: Any reason for changing the type from `auto` to `std::string`?

**Review by Yimura — CHANGES_REQUESTED** [review]

File: src/views/self/view_weapons.cpp (hunk 5 above)
Comment (TheGreenBandit) [reply src/views/self/view_weapons.cpp:223]: nope, just don't like auto

File: src/views/self/view_weapons.cpp (hunk 6 above)
Comment (TheGreenBandit) [reply src/views/self/view_weapons.cpp:228]: nope, just don't like auto

File: src/views/self/view_weapons.cpp (hunk 7 above)
Comment (TheGreenBandit) [reply src/views/self/view_weapons.cpp:337]: nope, just don't like auto

File: src/views/self/view_weapons.cpp (hunk 8 above)
Comment (TheGreenBandit) [reply src/views/self/view_weapons.cpp:287]: nope, just don't like auto

File: src/util/color.hpp (hunk 1 above)
Comment (TheGreenBandit) [reply src/util/color.hpp:14]: couldnt think of a different place since its being accessed by multiple files

Comment (Yimura) [reply src/util/color.hpp]: `core/data/color.hpp`?

File: src/backend/looped/weapons/paintgun.cpp (hunk 3 above)
Comment (tupoy-ya) [reply src/backend/looped/weapons/paintgun.cpp:50]: Can it?
It is using the custom weapon combo and no other members of it use commands.
https://github.com/YimMenu/YimMenu/blob/61bb60d1f430c932b3870478ba093d96d2895b6f/src/views/self/view_weapons.cpp#L107-L128

Comment (Yimura) [reply src/backend/looped/weapons/paintgun.cpp:50]: Probably not, I was a bit trigger happy with saying to use the looped command stuff.

For this specific example there's currently no way to implement it.

Comment (TheGreenBandit) [reply src/backend/looped/weapons/paintgun.cpp:50]: resolving this and tp gun as of comment above. Unless it is wanted a checkbox toggle which would be unnecessary.

**Review by Yimura — APPROVED** [review]
No further complaints.

File: src/backend/looped/weapons/paintgun.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+		if (g.weapons.custom_weapon == CustomWeapon::PAINT_GUN && (!g.self.custom_weapon_stop || WEAPON::IS_PED_ARMED(self::ped, 4 | 2)))
+		{
+			Vector3 c; entity::raycast(&c);
+
+			if (PAD::IS_DISABLED_CONTROL_PRESSED(0, (int)ControllerInputs::INPUT_AIM))
+			{
+				if (PAD::IS_DISABLED_CONTROL_PRESSED(0, (int)ControllerInputs::INPUT_ATTACK))
+				{
+					GRAPHICS::ADD_DECAL((int)DecalTypes::splatters_paint,
```
Comment (maybegreat48) [inline src/backend/looped/weapons/paintgun.cpp:26]: Is this networked?

Comment (TheGreenBandit) [reply src/backend/looped/weapons/paintgun.cpp:26]: No

**Pull request merged**